swap/src/database.rs

```diff
@@ -43,36 +47,56 @@ impl Display for Swap {
 
 #[derive(thiserror::Error, Debug, Clone, Copy, PartialEq)]
 #[error("Not in the role of Alice")]
-struct NotAlice;
+pub struct NotAlice;
 
 #[derive(thiserror::Error, Debug, Clone, Copy, PartialEq)]
 #[error("Not in the role of Bob")]
-struct NotBob;
+pub struct NotBob;
 
-impl Swap {
-    pub fn try_into_alice(self) -> Result<Alice> {
-        match self {
-            Swap::Alice(alice) => Ok(alice),
-            Swap::Bob(_) => bail!(NotAlice),
+
+impl TryFrom<Swap> for AliceState {
+    type Error =  NotAlice;
+
+    fn try_from(value: Swap) -> Result<Self, NotAlice> {
+        match value {
+            Swap::Alice(alice) => Ok(AliceState::from(alice)),
+            Swap::Bob(_) => Err(NotAlice),
         }
     }
+}
 
-    pub fn try_into_bob(self) -> Result<Bob> {
-        match self {
-            Swap::Bob(bob) => Ok(bob),
-            Swap::Alice(_) => bail!(NotBob),
+impl TryFrom<Swap> for BobState {
+    type Error = NotBob;
+
+    fn try_from(value: Swap) -> std::result::Result<Self, Self::Error> {
+        match value {
+            Swap::Bob(bob) => Ok(BobState::from(bob)),
+            Swap::Alice(_) => Err(NotBob),
         }
     }
 }
 
-pub struct Database {
+impl From<BobState> for Swap {
+    fn from(state: BobState) -> Self {
+        Self::Bob(Bob::from(state))
+    }
+}
+
+impl From<AliceState> for Swap {
+    fn from(state: AliceState) -> Self {
+        Self::Alice(Alice::from(state))
+    }
+}
+
+pub struct Database<T> where T: Into<Swap> + TryFrom<Swap>  {
```

Comment:
Try to avoid trait-bounds on structs but instead move them to the `impl` block there they are needed! If they are on the struct, you will have to repeat them everywhere where you use the struct which may ripple through your codebase.

Reply:
cool didnt know that